framework/src/utils/ADUtils.C

```diff
@@ -9,6 +9,10 @@
 
 #include "ADUtils.h"
 
+// We must include this first for the unity build to see some dataStore template specializations
+// before including DataIO.h through NonlinearSystemBase.h
```

Comment:
That is not good. Generally an indication that something smarter should be happening in the headers like some forward declarations

Reply:
I ll explore removing NLSystemBase.h before merging 

But it seems to me this type of ordering dependency is inherent to template specializations and the problem can happen again with including dataIO.h, and not just in this specific unity build case

Reply:
I can't forward declare in a source file so nothing I can do to remove NLSystemBase.h here

the more I think about it the more dataIO.h should include every other header that provides a template specialization of dataStore/dataLoad.
OR more reasonably,
all the template specialization declarations of dataStore/Load should be in dataIO.h, leveraging forward declarations in that class

Reply:
As long as a (declaration or definition) of the specialization is viewable in the TU ahead of the instantiation then everything is good. That’s why the ordering generally shouldn’t matter because the instantiating doesn’t occur until the body of the .C. I’m not really sure why unity should affect this or if it does, why this fixes it